### PR TITLE
added cephadm module support and test suite

### PR DESCRIPTION
**Pull request by sunilkumarn417**
- Added cephadm support and test suite.
- cephadm shell access support to run orchestration commands.

Test results: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1608799340416/result.html

Signed-off-by: sunilkumarn417 <sunnagar@redhat.com>

### CODE REVIEW

File: ceph/ceph.py [hunk 1]
```diff
@@ -1913,6 +1916,655 @@ def enable_ceph_mgr_restful(self):
...
+            #
+            #     for obj, cmd in radosgw_cmds.items():
+            #         try:
+            #             logger.info("creating {}...".format(obj))
+            #             self.shell(
+            #                 remote=self.installer,
+            #                 args=cmd,
+            #             )
+            #         except CommandFailed as warn:
+            #             logger.warning(warn)
```
Comment (rakeshgm) [inline ceph/ceph.py:2335]: why do we need these commented code here.

File: ceph/ceph.py [hunk 2]
```diff
@@ -1913,6 +1916,655 @@ def enable_ceph_mgr_restful(self):
...
+    #     """
+    #     try:
+    #         node.exec_command(
+    #             cmd='stat {disk} && sudo dd if={disk} of=/dev/null count=1'
+    #                 ' && ! mount | grep -q {disk}'.format(disk=disk)
+    #         )
+    #         return True
+    #     except:
+    #         logger.debug("get_scratch_devices: %s is in use" % disk)
+    #     return False
```
Comment (rakeshgm) [inline ceph/ceph.py:2377]: same as above, do we need this block of commented code.

**Comment by rakeshgm** [comment]
most of the code looks good to me. would request to address the above comments and we are good to go.

File: ceph/ceph.py (hunk 2 above)
Comment (sunilkumarn417) [reply ceph/ceph.py:2377]: Initially added to check the disk. we can eliminate this part currently.

File: ceph/ceph.py [hunk 3]
```diff
@@ -1913,6 +1916,606 @@ def enable_ceph_mgr_restful(self):
...
+                           for i in out if i.get('daemon_type') == daemon)
+
+            count = 0
+            for _id in ids:
+                if _id in daemons:
+                    status = daemons[_id]
+                    if status == 'running':
+                        count += 1
+
+            logger.info("{up_count}/{num} {daemon}s up....".format(num=len(ids),
```
Comment (psathyan) [inline ceph/ceph.py:2012]: Avoid using formatter with logging instead use `%`... read about it [doc](https://stackoverflow.com/questions/34619790/pylint-message-logging-format-interpolation)

File: ceph/ceph.py [hunk 4]
```diff
@@ -1913,6 +1916,606 @@ def enable_ceph_mgr_restful(self):
...
+                    )
+                    j = json.loads(out)
+                    if len(j['mons']) == num_mons:
+                        break
+                    sleep(sleep_time)
+                    timeout -= sleep_time
+
+        elif op == "purge":
+            # todo: Handle mon purge
+            pass
```
Comment (psathyan) [inline ceph/ceph.py:2210]: IMO, we could raise `NotImplementedError`

File: ceph/ceph.py [hunk 5]
```diff
@@ -1913,6 +1916,606 @@ def enable_ceph_mgr_restful(self):
...
+                file_name='/root/.ssh/authorized_keys', file_mode='a'
+            )
+            keys_file.write(ceph_pub_key)
+            keys_file.flush()
+            each_node.exec_command(
+                cmd="sudo chmod 0600 /root/.ssh/authorized_keys"
+            )
+
+
+class CephAdmin(object):
```
Comment (psathyan) [inline ceph/ceph.py:1957]: As we support Python3 only, we could leave it as `class CephAdmin`... you could read about it at [link](https://stackoverflow.com/questions/4015417/python-class-inherits-object)

Comment (psathyan) [reply ceph/ceph.py:1957]: FYI, there is nothing wrong in doing `class CephAdmin(object)`

**Review by psathyan — COMMENTED** [review]
IMHO, we should split `CephAdmin`  from `ceph.py`. This would ease our maintenance efforts and triage could be smoother. 

We would also be able to split them into smaller modules/objects based on the role. Allowing us to implement purge & create methods separately.

File: ceph/ceph.py (hunk 3 above)
Comment (rakeshgm) [reply ceph/ceph.py:2012]: though the advantage here is interpreting the strings when pylint is being used. but we have are not validating anything against pylint. if we have to formalise it we need to format the entire code base and emit these kinda messages. there are places we have `%` and there are places we have used `format` and there are places we have `f'{ }`

Comment (psathyan) [reply ceph/ceph.py:2012]: Agreed. I am also not in favor of refactoring... at some point in the future we would need to move to a standard based on consensus. @sunilkumarn417 you can chose to ignore this comment.

**Review by psathyan — APPROVED** [review]
Nice...+1

**Comment by sunilkumarn417** [comment]
latest result: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1609861771530/?C=M;O=A

```
All test logs located here: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1609861771530

TEST NAME                        TEST DESCRIPTION                                               DURATION                                  STATUS
install ceph pre-requisites      None                                                           0:05:35.819471                              Pass
cephadm deployment               cephadm cluster deployment                                     0:11:10.117310                              Pass
cephadm shell execution          cephadm commands execution                                     0:00:15.511207                              Pass
```

**Review by rakeshgm — APPROVED** [review]

**Pull request merged**